### PR TITLE
Awood/capacity usage

### PR DESCRIPTION
**Pull request by awood**

### CODE REVIEW

**Comment by awood** [comment]
@kahowell Not sure what's going wrong here.  SonarCloud is showing PR 141 (this one) as having passed.  Am I missing something?

**Comment by kahowell** [comment]
@awood so here's what happened:

1. Sonarcloud.io analyzed the branch itself and detected [one issue](https://sonarcloud.io/project/issues?branch=awood%2Fcapacity-usage&id=rhsm-subscriptions&resolved=false&sinceLeakPeriod=true) - [looking deeper](https://sonarcloud.io/project/settings?id=rhsm-subscriptions), "New Code Period" is currently set to detect new issues relative to the oldest commit within the last 30 days (super unintuitive, and flags things not changed in this PR...); I'm gonna change that to compare to master.
2. Sonarcloud.io analyzed the merged result and detected no issues, but then our webhook didn't get processed by Jenkins appropriately (have seen this a lot recently, I have an idea for a workaround I'll open a PR for shortly).

I'll kick off both again after the setting change; I think the checks will be good after that.

**Comment by kahowell** [comment]
@awood can you please rebase? Looking again, it looks like the branch check happens because your branch is missing a commit that fixed the issue (issue does not exist in master).

**Comment by awood** [comment]
> @awood can you please rebase?

Done!

File: src/main/java/org/candlepin/subscriptions/db/CustomizedSubscriptionCapacityRepository.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * Fragment interface defining methods for queries not handled well by Spring JPA's
+ * query methods DSL.
+ *
+ * See
+ * https://docs.spring.io/spring-data/jpa/docs/2.3.0.RELEASE/reference/html/#repositories.custom-implementations
+ */
+public interface CustomizedSubscriptionCapacityRepository {
+    @Transactional
+    List<SubscriptionCapacity> findByOwnerAndProductId(String ownerId, String productId,
+        ServiceLevel serviceLevel, Usage usage, OffsetDateTime begin, OffsetDateTime end);
```
[the remark this answers is not in the document]
Comment (awood) [reply src/main/java/org/candlepin/subscriptions/db/CustomizedSubscriptionCapacityRepository.java:42]: Yes.  That would help clear up the confusion a lot I think.

File: src/main/java/org/candlepin/subscriptions/db/CustomizedSubscriptionCapacityRepositoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        predicates.add(cb.equal(capacity.get("key").get("ownerId"), ownerId));
+        predicates.add(cb.equal(capacity.get("key").get("productId"), productId));
+
+        if (serviceLevel != null && serviceLevel != ServiceLevel.UNSPECIFIED) {
+            predicates.add(cb.equal(capacity.get("serviceLevel"), serviceLevel));
+        }
+
+        if (usage != null && usage != Usage.UNSPECIFIED) {
+            predicates.add(cb.equal(capacity.get("usage"), usage));
+        }
```
[the remark this answers is not in the document]
Comment (awood) [reply src/main/java/org/candlepin/subscriptions/db/CustomizedSubscriptionCapacityRepositoryImpl.java:75]: so ```if (usage != null && usage != Usage.ANY)```?

[the remark this answers is not in the document]
Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/db/CustomizedSubscriptionCapacityRepositoryImpl.java:75]: Precisely.

[the remark this answers is not in the document]
Comment (awood) [reply src/main/java/org/candlepin/subscriptions/db/CustomizedSubscriptionCapacityRepositoryImpl.java]: Even though UNSPECIFIED usages/service levels shouldn't get this far, I did a check for both just as a belt-and-suspenders approach.

**Pull request merged**